flocker/common/algebraic.py

```diff
@@ -0,0 +1,49 @@
+# -*- test-case-name: flocker.common.test.test_algebraic -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Invariants for algebraic data types.
+"""
+
+from pyrsistent import PClass, field
+
+__all__ = ["StateInvariant"]
+
+
+class StateInvariant(PClass):
+    """
+    An invariant that ensure the given object has a ``state`` attribute
+    in the given states, and that all the other specified attributes are
+    present if and only if the object is in one of the corresponding states.
+
+    :param set allowed_states: Set of allowed states.
+    :param dict expected_attributes: Dictionary mapping attribute names to
+        the set of states that attribute should be present in.
```

Comment:
I think it would be more clear to invert this dictionary, mapping states to the set of attributes that are allowed, since that's customary for ADTs in most other languages.